util/genrest/resttools/checkrestbody.go

```diff
@@ -0,0 +1,156 @@
+// limitations under the License.
+
+package resttools
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// CheckRESTBody verifies that any enum fields in message are properly represented in the JSON
+// payload carried by jsonReader: the fields must be either absent or have string values.
+func CheckRESTBody(jsonReader io.Reader, message protoreflect.Message) error {
+	jsonBytes, err := ioutil.ReadAll(jsonReader)
+	if err != nil {
+		return err
+	}
+	enumFields := GetEnumFields(message)
+	return CheckJSONEnumFields(jsonBytes, enumFields)
+}
+
+// CheckJSONEnumFields verifies that each of the fields listed in fieldsToCheck, presumably all
+// referring to enum fields, are encoded correctly in jsonBytes, meaning that the field is absent or
+// its value is a string. Each element of fieldsToCheck is a qualified proto field name represented
+// as a sequence of simple protoreflect.Name.
+func CheckJSONEnumFields(jsonBytes []byte, fieldsToCheck [][]protoreflect.Name) error {
+	// See See eg https://michaelheap.com/golang-encodedecode-arbitrary-json/
+
+	var payload map[string]interface{}
+	json.Unmarshal(jsonBytes, &payload)
+	badFields := []string{}
+	for _, fieldPath := range fieldsToCheck {
+		if field, ok := CheckEnum(payload, fieldPath); !ok {
+			badFields = append(badFields, field)
+		}
+	}
+	if len(badFields) > 0 {
+		return fmt.Errorf("badly transcoded enum values in fields: %v", badFields)
+	}
+	return nil
+}
+
+// CheckEnum verifies whether the field whose qualified name is captured in the elements of
+// fieldPath has a string value, if it exists, in the JSON representation captured by payload. This
+// returns the qualified field name (as present as it is in payload) as a single string, and a
+// boolean that is true only if either fieldPath is not present or if its value is a string. This
+// means that if fieldPath is a path to en enum field, the boolean will be false if the enum is
+// encoded in payloadusing a non-string representation.
+func CheckEnum(payload map[string]interface{}, fieldPath []protoreflect.Name) (fieldName string, ok bool) {
+	nameParts := []string{}
+	last := len(fieldPath) - 1
+	var value string
+	var found, isString bool
+	for idx, pathSegment := range fieldPath {
+		segment := string(pathSegment)
+		nameParts = append(nameParts, segment)
+
+		// TODO: For repeated fields, will need to recurse. Consider denoting repeated fields by appending a "*" to the pathSegment
+
+		if idx < last {
+			payload, found = payload[segment].(map[string]interface{})
+			if !found {
+				// Some elements of the field path are not populated
+				break
+			}
+			continue
+		}
+
+		if _, found = payload[segment]; found {
+			value, isString = payload[segment].(string)
+		}
+	}
+
+	fieldName = strings.Join(nameParts, ".")
+	if !found {
+		return fieldName, true
+	}
+
+	if !isString {
+		return fieldName, false
+	}
```

Comment:
Can you add some more supporting commentary in here? I got a little thrown by the loop control logic and how it determined what was being evaluated once the loop exited and got to these checks.

Reply:
Added some comments. WDYT?